core/src/main/java/org/infinispan/container/offheap/OffHeapDataContainer.java

```diff
@@ -327,30 +327,40 @@ protected void entryRetrieved(long entryAddress) {
     * key <b>must</b> be held before calling this method.
     * @param bucketHeadAddress the starting address of the address hash
     * @param key the key of the entry
+    * @param requireReturn whether this method is forced to return the entry removed (optimizations can be done if
+    *                      the entry is not needed)
     */
-   protected InternalCacheEntry<WrappedBytes, WrappedBytes> performRemove(long bucketHeadAddress, Object key) {
+   protected InternalCacheEntry<WrappedBytes, WrappedBytes> performRemove(long bucketHeadAddress, Object key,
+         boolean requireReturn) {
       WrappedByteArray wba = toWrapper(key);
       long prevAddress = 0;
       // We only use the head pointer for the first iteration
       long address = bucketHeadAddress;
+      InternalCacheEntry<WrappedBytes, WrappedBytes> ice = null;
 
       while (address != 0) {
          long nextAddress = offHeapEntryFactory.getNext(address);
-         InternalCacheEntry<WrappedBytes, WrappedBytes> ice = offHeapEntryFactory.fromMemory(address);
-         if (ice.getKey().equals(wba)) {
+         boolean removeThisAddress;
+         if (requireReturn) {
+            ice = offHeapEntryFactory.fromMemory(address);
```

Comment:
This could be done in if `removeThisAddress == true`.

Reply:
This one is a tough one imo. Do we want to optimize the miss case or the replace case?  My thought was that normally there would be just 1 address in the lookup, in which case the replace case would be hit more often.

The inverse is that if we optimize for replace (as is right now) it is more costly when a miss occurs though.

Reply:
I guess I am open to optimizing for the miss case, assuming that people have their address count not high enough.

Reply:
So `offHeapEntryFactory.equalsKey(address, wba)` is slower than `ice.getKey().equals(wba)`?

Reply:
Actually in ensureSize()` you already have the `addressToRemove` pointer, so you could compare addresses instead of keys in that case.

Reply:
No, my thought at the time (which as I think about it now was probably wrong). Would be that calling:

> offHeapEntryFactory.equalsKey(address, wba)
> ice = offHeapEntryFactory.fromMemory(address);
> return ice;

would be slower than 

> ice = offHeapEntryFactory.fromMemory(address);
> ice.getKey().equals(wba)
> return ice;

My mistaking thought at the time would be that CPU would be able to optimize the latter one, but thinking more the former actually has a better chance to be more optimized as the key could be smaller and fit in the cache between subsequent calls.

So I will change it as you suggested, Thanks! This will also fix the other comment.

Reply:
>Actually in ensureSize()you already have theaddressToRemove` pointer, so you could compare addresses instead of keys in that case.

Hrmm that is a good point. You mean to add an additional argument to `performRemove` that takes the `bucketHeadAddress` and the `addressToRemove` and then we don't even have to do the equals check at all. That is even way better :+1:

Reply:
Ok, I thought you saw something in the profiler that suggests HotSpot can't optimize `equalsKey`.

Reply:
Actually this can apply to all the write operations (evict, remove or put/compute that replace an entry 👍